grails-app/services/org/pih/warehouse/core/DocumentTemplateService.groovy

```diff
@@ -0,0 +1,221 @@
+* You must not remove this notice, or any other, from this software.
+**/
+package org.pih.warehouse.core
+
+import fr.opensagres.xdocreport.converter.ConverterTypeTo
+import fr.opensagres.xdocreport.converter.ConverterTypeVia
+import fr.opensagres.xdocreport.converter.Options
+import fr.opensagres.xdocreport.document.IXDocReport
+import fr.opensagres.xdocreport.document.registry.XDocReportRegistry
+import fr.opensagres.xdocreport.template.IContext
+import fr.opensagres.xdocreport.template.TemplateEngineKind
+import fr.opensagres.xdocreport.template.formatter.FieldsMetadata
+import groovy.text.Template
+import org.codehaus.groovy.grails.web.pages.GroovyPagesTemplateEngine
+import org.pih.warehouse.order.Order
+import org.pih.warehouse.order.OrderAdjustment
+import org.pih.warehouse.order.OrderItem
+import org.pih.warehouse.shipping.Shipment
+import org.pih.warehouse.shipping.ShipmentItem
+
+class DocumentTemplateService {
+
+    boolean transactional = true
+    GroovyPagesTemplateEngine groovyPagesTemplateEngine
+
+    def renderGroovyServerPageDocumentTemplate(Document documentTemplate, Map model) {
+        StringWriter output = new StringWriter()
+        String templateContents = new String(documentTemplate.fileContents)
+        Template template = groovyPagesTemplateEngine.createTemplate(templateContents, documentTemplate.name)
+        template.make(model).writeTo(output)
+        return output.toString()
+    }
+
+    def renderOrderDocumentTemplate(Document documentTemplate, Order orderInstance, ConverterTypeTo targetDocumentType, OutputStream outputStream) {
+
+        Boolean isVelocityTemplate = documentTemplate.filename.contains(".vm") || documentTemplate.filename.contains(".vtl")
+
+        TemplateEngineKind templateEngineKind = isVelocityTemplate ?
+                TemplateEngineKind.Velocity : TemplateEngineKind.Freemarker
+
+        InputStream inputStream = new ByteArrayInputStream(documentTemplate.fileContents)
+        IXDocReport report = XDocReportRegistry.getRegistry().loadReport(inputStream, templateEngineKind);
+
+        // FIXME Need a better way to handle this generically (consider using config + dataService)
+        IContext context = orderInstance ? createOrderContext(report, orderInstance) : report.createContext();
+
+        ConverterTypeVia sourceDocumentType = report.kind == "DOCX" ? ConverterTypeVia.DOCX4J :
+                report.getKind() == "ODT" ? ConverterTypeVia.ODFDOM : ConverterTypeVia.XWPF
+
+        // Convert to the target type
+        if (targetDocumentType && sourceDocumentType) {
+            Options options = Options.getTo(targetDocumentType).via(sourceDocumentType);
+            report.convert(context, options, outputStream);
+        }
+        // Process document in place
+        else {
+            report.process(context, outputStream)
+        }
+        log.info "Report dump: " + report.dump()
+
+    }
+
+    def createOrderContext(IXDocReport report, Order orderInstance) {
```

Comment:
I'm ok with an exception being thrown here if report is null. I'm guessing that's not possible because there would be an exception when loading the report so it would not return a null. And orderInstance will not be null as we handle that case in the controller action.